---
title: "calculate_eigenvectors_wecare_only_jan2017"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 01Mar2017

# Summary

Overall, eigenvectors are calculated for 3 datasets:  

- wecare-only (all variants and common variants)
- wecare-nfe (all variants and common variants)  
- wecare-nfe-50kgen (overlapped-variants and overlapped-common-variants)  

This script deals with wecare-only dataset.  
Additionally it exports wecare-only data to text files in eigenstrat format. 

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
setwd("/scratch/medgen/scripts/wecare_stat_01.17/scripts")
interim_data_folder <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data"

# Required libraries
library(dplyr)
library(ggplot2)
library(plotly)

```

# load_data

```{r load_data}

load(paste(interim_data_folder, "r04_filter_cases_jan2017.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(wecare_nfe_variants.df)
colnames(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

# Check consistency of colnames and rownames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))

```

# function_to_calculate_eigenvectors

Implements procedure described by Price et al 2006 (PMID: 16862161)

```{r function_to_calculate_eigenvectors}

normalise_and_calculate_eigenvectors.udf <- function(x) {
  
  # --- Center and normalise variants (rows) --- #
  
  # Center by mean
  avg.rows <- apply(x, 1, mean, na.rm=TRUE)
  x.c <- x - avg.rows
  
  # Normalise by sqrt(p(1-p)) where p~"posterior estimate of unobserved allele frequency"
  # This is motivated by the fact that genetic drift per generation is proportional to this normalisation value (Patterson 2006)
  # Also this makes each column to have same variance
  # 
  p.fnc <- function(x) (1 + sum(x, na.rm=TRUE)) / (2 + 2 * sum(!is.na(x)))
  p <- apply(x, 1, p.fnc)
  eaf <- sqrt(p*(1-p))
  x.cn <- x.c/eaf
  
  # Substitute NAs to zeros
  0 -> x.cn[is.na(x)]
  
  # --- Calculate eigenvectors of covariance matrix of cases --- #
  
  cov.mx <- cov(x.cn)
  eig <- eigen(cov.mx) # eigenvectors in columns
  
  return(eig)

}

```

# make_wecare_only_subset_of_data

```{r make_wecare_only_subset_of_data}

# Remove nfe cases
wecare_cases <- as.vector(wecare_nfe_phenotypes.df[wecare_nfe_phenotypes.df$cc!=-1,"wes_id"])
length(wecare_cases) # 480

wecare_genotypes.mx <- wecare_nfe_genotypes.mx[,wecare_cases]
dim(wecare_genotypes.mx) # 275,516 x 480
wecare_genotypes.mx[1:5,1:5]

wecare_phenotypes.df <- wecare_nfe_phenotypes.df[wecare_cases,]
dim(wecare_phenotypes.df) # 480 x 28
wecare_phenotypes.df[1:5,1:5]

# Clean-up
rm(wecare_cases)

```

# remove_variants_with_uniform_genotypes_accross_all_wecare_samples
Remove 50,141 variants: 275,516 -> 225,375

```{r remove_variants_with_uniform_genotypes_accross_all_wecare_samples}

# Check that there is no all-NA variants
non_NA_count.udf <- function(x){sum(!is.na(x))}
all_NA <- apply(wecare_genotypes.mx, 1, non_NA_count.udf) == 0
sum(all_NA) # 0

# Function to detect uniform numeric vector
uniform_vector.udf <- function(x){
  if(min(x, na.rm=TRUE) == max(x, na.rm=TRUE)){return(TRUE)} else {return(FALSE)}}

# Variants with uniform genotypes accross all samples 
uniform_genotypes <- apply(wecare_genotypes.mx, 1, uniform_vector.udf)
summary(uniform_genotypes)
sum(uniform_genotypes) # 50,141

# Remove variants with uniform genotypes accross all samples
wecare_genotypes.mx <- wecare_genotypes.mx[!uniform_genotypes,]
dim(wecare_genotypes.mx) # 225375    480
wecare_genotypes.mx[1:5,1:5]

wecare_variants.df <- wecare_nfe_variants.df[!uniform_genotypes,]
dim(wecare_variants.df) # 225375     23
wecare_variants.df[1:5,1:5]

wecare_kgen.df <- wecare_nfe_kgen.df[!uniform_genotypes,]
dim(wecare_kgen.df) # 225375      9
wecare_kgen.df[1:5,1:5]

wecare_exac.df <- wecare_nfe_exac.df[!uniform_genotypes,]
dim(wecare_exac.df) # 225375     48
wecare_exac.df[1:5,1:5]

# Clean-up
rm(non_NA_count.udf, all_NA, uniform_vector.udf, uniform_genotypes)

```

# remove_wecare_nfe_tables

```{r remove_wecare_nfe_tables}

rm(wecare_nfe_exac.df, wecare_nfe_kgen.df, wecare_nfe_variants.df, wecare_nfe_genotypes.mx, wecare_nfe_phenotypes.df)

```

# calculate_and_plot_eigenvectors

```{r calculate_and_plot_eigenvectors}

# --- Calculate eigenvectors --- #

wecare.eigen <- normalise_and_calculate_eigenvectors.udf(wecare_genotypes.mx)

wecare.eigenvectors.df <- as.data.frame(wecare.eigen$vectors) # eigenvectors in columns
wecare.eigenvalues <- wecare.eigen$values

# --- Prepare data for plotting --- #

# Prepare colour scale
colours <- c("UBC" = "BLUE", "CBC" = "RED")
userColourScale <- scale_colour_manual(values=colours)

# Prepare cases lables
cases_labels <- as.vector(wecare_phenotypes.df$cc)
"CBC" -> cases_labels[cases_labels==1]
"UBC" -> cases_labels[cases_labels==0]

# Prepare cases IDs
cases_IDs <- as.vector(wecare_phenotypes.df$wes_id)

# make the dataframe
data2plot.df <- cbind(cases_IDs, cases_labels, wecare.eigenvectors.df[,1:3])
colnames(data2plot.df) <- c("wes_id", "group", "ev1", "ev2", "ev3")

# --- Plot eig1 vs eig2 --- #

g <- ggplot(data2plot.df, aes(ev1, ev2)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare all variants<br>(225,375 x 480)", x ="eigenvector1", y = "eigenvector2") +
  userColourScale

ggplotly(g)

# --- Plot eig2 vs eig3 --- #

g <- ggplot(data2plot.df, aes(ev2, ev3)) +
  geom_point(aes(colour=group, fill=group, text = wes_id)) + 
  labs(title="wecare all variants<br>(225,375 x 480)", x ="eigenvector2", y = "eigenvector3") +
  userColourScale

ggplotly(g)

# --- Clean-up --- #

rm(wecare.eigenvectors.df, wecare.eigenvalues, g, data2plot.df, 
   cases_IDs, cases_labels, colours, userColourScale, 
   normalise_and_calculate_eigenvectors.udf)

```

# calculate_outliers

Using 6 standard deviations in 5 eigenvectors

```{r calculate_outliers}

wecare.eigenvectors.mx <- wecare.eigen$vectors

ev1 <- wecare.eigenvectors.mx[,1]
ev1.positive_outliers <- ev1 > mean(ev1) + 6 * sd(ev1)
ev1.negative_outliers <- ev1 < mean(ev1) - 6 * sd(ev1)
sum(ev1.positive_outliers)
sum(ev1.negative_outliers)

wecare_phenotypes.df$wes_id[ev1.positive_outliers]
wecare_phenotypes.df$wes_id[ev1.negative_outliers]

ev2 <- wecare.eigenvectors.mx[,2]
ev2.positive_outliers <- ev2 > mean(ev2) + 6 * sd(ev2)
ev2.negative_outliers <- ev2 < mean(ev2) - 6 * sd(ev2)
sum(ev2.positive_outliers)
sum(ev2.negative_outliers)

wecare_phenotypes.df$wes_id[ev2.positive_outliers]
wecare_phenotypes.df$wes_id[ev2.negative_outliers]

ev3 <- wecare.eigenvectors.mx[,3]
ev3.positive_outliers <- ev3 > mean(ev3) + 6 * sd(ev3)
ev3.negative_outliers <- ev3 < mean(ev3) - 6 * sd(ev3)
sum(ev3.positive_outliers)
sum(ev3.negative_outliers)

wecare_phenotypes.df$wes_id[ev3.positive_outliers]
wecare_phenotypes.df$wes_id[ev3.negative_outliers]

ev4 <- wecare.eigenvectors.mx[,4]
ev4.positive_outliers <- ev4 > mean(ev4) + 6 * sd(ev4)
ev4.negative_outliers <- ev4 < mean(ev4) - 6 * sd(ev4)
sum(ev4.positive_outliers)
sum(ev4.negative_outliers)

wecare_phenotypes.df$wes_id[ev4.positive_outliers]
wecare_phenotypes.df$wes_id[ev4.negative_outliers]

ev5 <- wecare.eigenvectors.mx[,5]
ev5.positive_outliers <- ev5 > mean(ev5) + 6 * sd(ev5)
ev5.negative_outliers <- ev5 < mean(ev5) - 6 * sd(ev5)
sum(ev5.positive_outliers)
sum(ev5.negative_outliers)

wecare_phenotypes.df$wes_id[ev5.positive_outliers]
wecare_phenotypes.df$wes_id[ev5.negative_outliers]

# Clean-up
rm(wecare.eigenvectors.mx, ev1, ev1.positive_outliers, ev1.negative_outliers, 
   ev2, ev2.positive_outliers, ev2.negative_outliers, ev3, ev3.positive_outliers, ev3.negative_outliers,
   ev4, ev4.positive_outliers, ev4.negative_outliers, ev5, ev5.positive_outliers, ev5.negative_outliers)

```

# export_to_EIGENSTRAT_format

```{r export_to_EIGENSTRAT_format}

# --- make snp file --- #

# Extract data from variants table
wecare_snp.df <- wecare_variants.df %>% 
  mutate(morgan = POS / 1000000) %>%  
  select(SplitVarID, CHROM, morgan, POS, REF, ALT)

# Recode CRHOM data
summary(wecare_snp.df$CHROM)

str(wecare_snp.df)
wecare_snp.df$CHROM <- as.vector(wecare_snp.df$CHROM)
str(wecare_snp.df)

"23" -> wecare_snp.df[wecare_snp.df$CHROM == "X", "CHROM"]
"24" -> wecare_snp.df[wecare_snp.df$CHROM == "Y", "CHROM"]
"90" -> wecare_snp.df[wecare_snp.df$CHROM == "MT", "CHROM"]

wecare_snp.df$CHROM <- as.factor(wecare_snp.df$CHROM)
summary(wecare_snp.df$CHROM)

# Write file (tab-separated)
write.table(wecare_snp.df, 
            paste(interim_data_folder, "wecare.snp", sep="/"), 
            quote=FALSE, sep="\t", 
            row.names=FALSE, col.names=FALSE) 

# --- make geno file --- #

# Get data from genotypes matrix
wecare_geno.mx <- wecare_genotypes.mx

# Recode NA
9 -> wecare_geno.mx[is.na(wecare_geno.mx)] # NA is coded as 9

# Write file (no delimiters!)
write.table(wecare_geno.mx, 
            paste(interim_data_folder, "wecare.geno", sep="/"),             
            quote=FALSE, sep="", 
            row.names=FALSE, col.names=FALSE)

# --- make ind file --- #

# Samples column
samples <- as.vector(wecare_phenotypes.df$wes_id)

# Gender column
gender <- rep("F", length(samples)) # all females

# Population/Group column
group <- as.vector(wecare_phenotypes.df$cc)
"UBC" -> group[group == 0]
"CBC" -> group[group == 1]

# Get together
wecare_ind.mx <- cbind(samples, gender, group)

# Write file (tab-separated) 
write.table(wecare_ind.mx, 
            paste(interim_data_folder, "wecare.ind", sep="/"),
            quote=FALSE, sep="\t", 
            row.names=FALSE, col.names=FALSE)

# Clean-up
rm(wecare_snp.df, wecare_geno.mx, wecare_ind.mx, samples, gender, group)

```

# data_summary

```{r data_summary}

ls()

dim(wecare_genotypes.mx)
class(wecare_genotypes.mx)
wecare_genotypes.mx[1:5,1:5]

dim(wecare_phenotypes.df)
str(wecare_phenotypes.df)
wecare_phenotypes.df[1:5,1:5]

dim(wecare_variants.df)
colnames(wecare_variants.df)
wecare_variants.df[1:5,1:5]

dim(wecare_kgen.df)
colnames(wecare_kgen.df)
wecare_kgen.df[1:5,1:5]

dim(wecare_exac.df)
colnames(wecare_exac.df)
wecare_exac.df[1:5,1:5]

str(wecare.eigen)

sum(colnames(wecare_genotypes.mx) != rownames(wecare_phenotypes.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_variants.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_kgen.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_exac.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r05a_calculate_egenvectors_wecare_only_jan2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
