---
title: "calculate_eigenvectors_wecare_nfe_jan2017"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 01Mar2017

# Summary

Overall, eigenvectors are calculated for 3 datasets:  

- wecare-only (all variants and common variants)
- wecare-nfe (all variants and common variants)  
- wecare-nfe-50kgen (overlapped-variants and overlapped-common-variants)  

This script deals with wecare-nfe dataset 

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
setwd("/scratch/medgen/scripts/wecare_stat_01.17/scripts")
interim_data_folder <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data"

# Required libraries
library(dplyr)
library(ggplot2)
library(plotly)

```

# load_data

```{r load_data}

load(paste(interim_data_folder, "r04_filter_cases_jan2017.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(wecare_nfe_variants.df)
colnames(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

# Check consistency of colnames and rownames
sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))

```

# function_to_calculate_eigenvectors

Implements procedure described by Price et al 2006 (PMID: 16862161)

```{r function_to_calculate_eigenvectors}

normalise_and_calculate_eigenvectors.udf <- function(x) {
  
  # --- Center and normalise variants (rows) --- #
  
  # Center by mean
  avg.rows <- apply(x, 1, mean, na.rm=TRUE)
  x.c <- x - avg.rows
  
  # Normalise by sqrt(p(1-p)) where p~"posterior estimate of unobserved allele frequency"
  # This is motivated by the fact that genetic drift per generation is proportional to this normalisation value (Patterson 2006)
  # Also this makes each column to have same variance
  # 
  p.fnc <- function(x) (1 + sum(x, na.rm=TRUE)) / (2 + 2 * sum(!is.na(x)))
  p <- apply(x, 1, p.fnc)
  eaf <- sqrt(p*(1-p))
  x.cn <- x.c/eaf
  
  # Substitute NAs to zeros
  0 -> x.cn[is.na(x)]
  
  # --- Calculate eigenvectors of covariance matrix of cases --- #
  
  cov.mx <- cov(x.cn)
  eig <- eigen(cov.mx) # eigenvectors in columns
  
  return(eig)

}

```

# calculate_and_plot_eigenvectors_for_wecare_nfe

```{r calculate_and_plot_eigenvectors_for_wecare_nfe}

# --- Calculate eigenvectors --- #

wecare_nfe.eigen <- normalise_and_calculate_eigenvectors.udf(wecare_nfe_genotypes.mx)

wecare_nfe.eigenvectors.df <- as.data.frame(wecare_nfe.eigen$vectors) # eigenvectors in columns
wecare_nfe.eigenvalues <- wecare_nfe.eigen$values

# --- Prepare data for plotting --- #

# Prepare cases lables
cases_labels <- as.vector(wecare_nfe_phenotypes.df$cc)
"NFE" -> cases_labels[cases_labels==-1]
"UBC" -> cases_labels[cases_labels==0]
"CBC" -> cases_labels[cases_labels==1]

summary(as.factor(cases_labels))

# Cases IDs (for interactive plot)
cases_IDs <- as.vector(wecare_nfe_phenotypes.df$wes_id)

# Prepare colour scale
colours <- c("NFE" = "BLUE", "UBC" = "PINK", "CBC" = "RED")
userColourScale <- scale_colour_manual(values=colours)

# Data frame to plot
data2plot.df <- cbind(cases_IDs, cases_labels, wecare_nfe.eigenvectors.df[,1:3])

# --- Plot eig1 vs eig2 --- #

g <- ggplot(data2plot.df, aes(-V1,V2)) +
  geom_point(aes(colour=cases_labels, text=cases_IDs)) + 
  labs(title="wecare-nfe<br>all variants (275,516 x 678)", x="-eigenvector1", y="eigenvector2") +
  userColourScale

ggplotly(g)

# --- Clean-up --- #

rm(wecare_nfe.eigenvectors.df, wecare_nfe.eigenvalues, g, data2plot.df, 
   cases_IDs, cases_labels, colours, userColourScale, 
   normalise_and_calculate_eigenvectors.udf)

```

# calculate_eigenvector_outliers_for_wecare_nfe

Using 6 standard deviations in 5 eigenvectors

```{r calculate_eigenvector_outliers_for_wecare_nfe}

wecare_nfe_all_variants_eigenvectors.mx <- wecare_nfe.eigen$vectors

ev1 <- wecare_nfe_all_variants_eigenvectors.mx[,1]
ev1.positive_outliers <- ev1 > mean(ev1) + 6 * sd(ev1)
ev1.negative_outliers <- ev1 < mean(ev1) - 6 * sd(ev1)
sum(ev1.positive_outliers)
sum(ev1.negative_outliers)

wecare_nfe_phenotypes.df$wes_id[ev1.positive_outliers]
wecare_nfe_phenotypes.df$wes_id[ev1.negative_outliers]

ev2 <- wecare_nfe_all_variants_eigenvectors.mx[,2]
ev2.positive_outliers <- ev2 > mean(ev2) + 6 * sd(ev2)
ev2.negative_outliers <- ev2 < mean(ev2) - 6 * sd(ev2)
sum(ev2.positive_outliers)
sum(ev2.negative_outliers)

wecare_nfe_phenotypes.df$wes_id[ev2.positive_outliers]
wecare_nfe_phenotypes.df$wes_id[ev2.negative_outliers]

ev3 <- wecare_nfe_all_variants_eigenvectors.mx[,3]
ev3.positive_outliers <- ev3 > mean(ev3) + 6 * sd(ev3)
ev3.negative_outliers <- ev3 < mean(ev3) - 6 * sd(ev3)
sum(ev3.positive_outliers)
sum(ev3.negative_outliers)

wecare_nfe_phenotypes.df$wes_id[ev3.positive_outliers]
wecare_nfe_phenotypes.df$wes_id[ev3.negative_outliers]

ev4 <- wecare_nfe_all_variants_eigenvectors.mx[,4]
ev4.positive_outliers <- ev4 > mean(ev4) + 6 * sd(ev4)
ev4.negative_outliers <- ev4 < mean(ev4) - 6 * sd(ev4)
sum(ev4.positive_outliers)
sum(ev4.negative_outliers)

wecare_nfe_phenotypes.df$wes_id[ev4.positive_outliers]
wecare_nfe_phenotypes.df$wes_id[ev4.negative_outliers]

ev5 <- wecare_nfe_all_variants_eigenvectors.mx[,5]
ev5.positive_outliers <- ev5 > mean(ev5) + 6 * sd(ev5)
ev5.negative_outliers <- ev5 < mean(ev5) - 6 * sd(ev5)
sum(ev5.positive_outliers)
sum(ev5.negative_outliers)

wecare_nfe_phenotypes.df$wes_id[ev5.positive_outliers]
wecare_nfe_phenotypes.df$wes_id[ev5.negative_outliers]

# Clean-up
rm(wecare_nfe_all_variants_eigenvectors.mx, ev1, ev1.positive_outliers, ev1.negative_outliers, 
   ev2, ev2.positive_outliers, ev2.negative_outliers, ev3, ev3.positive_outliers, ev3.negative_outliers,
   ev4, ev4.positive_outliers, ev4.negative_outliers, ev5, ev5.positive_outliers, ev5.negative_outliers)

```

# export_wecare_nfe_to_EIGENSTRAT_format

Omitted

# data_summary

```{r data_summary}

ls()

# wecare nfe all vars

dim(wecare_nfe_genotypes.mx)
class(wecare_nfe_genotypes.mx)
wecare_nfe_genotypes.mx[1:5,1:5]

dim(wecare_nfe_phenotypes.df)
str(wecare_nfe_phenotypes.df)
wecare_nfe_phenotypes.df[1:5,1:5]

dim(wecare_nfe_variants.df)
colnames(wecare_nfe_variants.df)
wecare_nfe_variants.df[1:5,1:5]

dim(wecare_nfe_kgen.df)
colnames(wecare_nfe_kgen.df)
wecare_nfe_kgen.df[1:5,1:5]

dim(wecare_nfe_exac.df)
colnames(wecare_nfe_exac.df)
wecare_nfe_exac.df[1:5,1:5]

str(wecare_nfe.eigen)

sum(colnames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_phenotypes.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_variants.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_kgen.df))
sum(rownames(wecare_nfe_genotypes.mx) != rownames(wecare_nfe_exac.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r05b_calculate_egenvectors_wecare_nfe_jan2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
