---
title: "compare_eigenvectors_wecare_only_jan2017"
output: html_document
---

started: Alexey Larionov, 30Jan2017  
last updated: Alexey Larionov, 10Feb2017

# Summary

1) Compare eigenvectors calculated by R and by Eigenstrat PCA  
  - using wecare only dataset w/o removing outliers (outliermode 2)  
  - using Price 2006 normalising (altnormstyle NO)  

2) Compare WES and GWAS eigenvectors 

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
setwd("/scratch/medgen/scripts/wecare_stat_01.17/scripts")
interim_data_folder <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data"

# Required libraries
library(dplyr)
library(ggplot2)
library(plotly)

```

# read_r_data

```{r read_r_data}

# --- eigenvectors --- #

load(paste(interim_data_folder, "r05a_calculate_egenvectors_wecare_only_jan2017.RData", sep="/"))

rm(wecare_exac.df, wecare_genotypes.mx, wecare_kgen.df, wecare_nfe_exac.df, wecare_nfe_genotypes.mx, wecare_nfe_kgen.df, wecare_nfe_phenotypes.df, wecare_nfe_variants.df, wecare_variants.df, 
common_wecare_exac.df, common_wecare_genotypes.mx, common_wecare_kgen.df, common_wecare_variants.df)

# eigenvectors for all variants
r_evectors.df <- as.data.frame(wecare_eigen$vectors[,1:10])

colnames(r_evectors.df) <- c("r_ev1", "r_ev2", "r_ev3", "r_ev4", "r_ev5", "r_ev6", "r_ev7", "r_ev8", "r_ev9", "r_ev10")
str(r_evectors.df)

# eigenvectors for common variants
common_r_evectors.df <- as.data.frame(common_wecare_eigen$vectors[,1:10])

colnames(common_r_evectors.df) <- c("r_ev1", "r_ev2", "r_ev3", "r_ev4", "r_ev5", "r_ev6", "r_ev7", "r_ev8", "r_ev9", "r_ev10")
str(common_r_evectors.df)

# Common sence checks (using properties of eigenvectors)
apply(r_evectors.df,2,sum) # should be close to zeros
apply(r_evectors.df,2,function(x){sum(x^2)}) # should be close to ones

# --- eigenvalues --- #

# all variants
r_evalues.df <- as.data.frame(wecare_eigen$values)
colnames(r_evalues.df) <- "evals"
plot(r_evalues.df$evals, main="Eigenvalues (WECARE, all variants, R-script)")

# common variants
common_r_evalues.df <- as.data.frame(common_wecare_eigen$values)
colnames(common_r_evalues.df) <- "evals"
plot(common_r_evalues.df$evals, main="Eigenvalues (WECARE, common variants, R-script)")

```

# read_es_data

```{r read_es_data}

# --- eigenvectors --- #

es_evectors_file="/scratch/medgen/scripts/wecare_stat_01.17/interim_data/u01_wecare_all_variants_eigenstrat_no_outliers/u01_wecare_all_variants_eigenstrat_no_outliers.evec"

common_es_evectors_file="/scratch/medgen/scripts/wecare_stat_01.17/interim_data/u01_wecare_common_variants_eigenstrat_no_outliers/u01_wecare_common_variants_eigenstrat_no_outliers.evec"

# eigenvectors for all variants
es_evectors.df <- read.table(es_evectors_file)
common_es_evectors.df <- read.table(common_es_evectors_file)

colnames(es_evectors.df) <- c("wes_id", "es_ev1", "es_ev2", "es_ev3", "es_ev4", "es_ev5", "es_ev6", "es_ev7", "es_ev8", "es_ev9", "es_ev10", "group")

colnames(common_es_evectors.df) <- c("wes_id", "es_ev1", "es_ev2", "es_ev3", "es_ev4", "es_ev5", "es_ev6", "es_ev7", "es_ev8", "es_ev9", "es_ev10", "group")

str(es_evectors.df)
str(common_es_evectors.df)

# Common sence checks (using properties of eigenvectors)
apply(es_evectors.df[,2:11],2,sum) # should be close to zeros
apply(es_evectors.df[,2:11],2,function(x){sum(x^2)}) # should be close to ones
cor.test(as.numeric(es_evectors.df$es_ev1), as.numeric(es_evectors.df$es_ev2)) # should be orthogonal

apply(common_es_evectors.df[,2:11],2,sum) # should be close to zeros
apply(common_es_evectors.df[,2:11],2,function(x){sum(x^2)}) # should be close to ones
cor.test(as.numeric(common_es_evectors.df$es_ev1), as.numeric(common_es_evectors.df$es_ev2)) # should be orthogonal

# --- eigenvalues --- #

es_evalues_file <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data/u01_wecare_all_variants_eigenstrat_no_outliers/u01_wecare_all_variants_eigenstrat_no_outliers.eval"

common_es_evalues_file <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data/u01_wecare_common_variants_eigenstrat_no_outliers/u01_wecare_common_variants_eigenstrat_no_outliers.eval"

es_evalues.df <- read.table(es_evalues_file)
common_es_evalues.df <- read.table(common_es_evalues_file)

colnames(es_evalues.df) <- "evals"
str(es_evalues.df)
plot(es_evalues.df$evals, main="Eigenvalues (WECARE, all variants, EIGENSTRAT)")

colnames(common_es_evalues.df) <- "evals"
str(common_es_evalues.df)
plot(common_es_evalues.df$evals, main="Eigenvalues (WECARE, common variants, EIGENSTRAT)")

# --- Clean-up --- #
rm(es_evectors_file, es_evalues_file, common_es_evectors_file, common_es_evalues_file)

```

# compare_evalues

```{r compare_evalues}

# All variants
plot(es_evalues.df$evals, r_evalues.df$evals, main="eigenvalues: R vs EIGENSTRAT (all variants)")
abline(c(0,0),c(1,1), col="red")

# Common variants
plot(common_es_evalues.df$evals, common_r_evalues.df$evals, main="eigenvalues: R vs EIGENSTRAT (common variants)")
abline(c(0,0),c(1,1), col="red")

```

# compare_evectors_all_variants

```{r compare_evectors_all_variants}

# Prepare dataframe
joined_evectors.df <- cbind(es_evectors.df, r_evectors.df)

# Prepare colour scale
colours <- c("UBC" = "BLUE", "CBC" = "RED")
userColourScale <- scale_colour_manual(values=colours)

# --- 1st eigenvectors --- #

g <- ggplot(joined_evectors.df, aes(es_ev1, r_ev1)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="1st eigenvectors: R vs EIGENSTRAT<br>WECARE, all variants") +
  geom_abline(colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(joined_evectors.df$es_ev1), as.numeric(joined_evectors.df$r_ev1))

# --- 2nd eigenvectors --- #

g <- ggplot(joined_evectors.df, aes(es_ev2, r_ev2)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="2nd eigenvectors: R vs EIGENSTRAT<br>WECARE, all variants") +
  geom_abline(colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(joined_evectors.df$es_ev2), as.numeric(joined_evectors.df$r_ev2))

# --- 3rd eigenvectors --- #

g <- ggplot(joined_evectors.df, aes(es_ev3, r_ev3)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="3rd eigenvectors: R vs EIGENSTRAT<br>WECARE, all variants") +
  geom_abline(slope = -1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(joined_evectors.df$es_ev3), as.numeric(joined_evectors.df$r_ev3))

# Clean-up
rm(joined_evectors.df)

```

# compare_evectors_common_variants

```{r compare_evectors_common_variants}

# Prepare dataframe
common_joined_evectors.df <- cbind(common_es_evectors.df, common_r_evectors.df)

# --- 1st eigenvectors --- #

g <- ggplot(common_joined_evectors.df, aes(es_ev1, r_ev1)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="1st eigenvectors: R vs EIGENSTRAT<br>WECARE, common variants") +
  geom_abline(colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(common_joined_evectors.df$es_ev1), as.numeric(common_joined_evectors.df$r_ev1))

# --- 2nd eigenvectors --- #

g <- ggplot(common_joined_evectors.df, aes(es_ev2, r_ev2)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="2nd eigenvectors: R vs EIGENSTRAT<br>WECARE, common variants") +
  geom_abline(colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(common_joined_evectors.df$es_ev2), as.numeric(common_joined_evectors.df$r_ev2))

# --- 3rd eigenvectors --- #

g <- ggplot(common_joined_evectors.df, aes(es_ev3, r_ev3)) +
  geom_point(aes(colour=group, fill=group, text=wes_id)) + 
  labs(title="3rd eigenvectors: R vs EIGENSTRAT<br>WECARE, common variants") +
  geom_abline(slope = -1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(as.numeric(common_joined_evectors.df$es_ev3), as.numeric(common_joined_evectors.df$r_ev3))

# Clean-up
rm(common_joined_evectors.df)

```

# compare_WES_and_GWAS_eigenvectors_EIGENSTRAT_all_variants

```{r compare_WES_and_GWAS_eigenvectors_EIGENSTRAT_all_variants}

# Prepare plot data

data2plot.df <- as.data.frame(cbind(
  es_evectors.df$es_ev1,
  es_evectors.df$es_ev2, 
  es_evectors.df$es_ev3, 
  wecare_phenotypes.df$eig1_gwas,
  wecare_phenotypes.df$eig2_gwas,
  wecare_phenotypes.df$eig3_gwas,
  wecare_phenotypes.df$cc))

sum(es_evectors.df$wes_id != wecare_phenotypes.df$wes_id)

data2plot.df <- cbind(es_evectors.df$wes_id, data2plot.df)

colnames(data2plot.df) <- c("wes_id", "wes_ev1", "wes_ev2", "wes_ev3", 
                            "gwas_ev1", "gwas_ev2", "gwas_ev3", "group")

"UBC" -> data2plot.df[data2plot.df$group == 0, "group"] # "UBC" = 0 = "BLUE"
"CBC" -> data2plot.df[data2plot.df$group == 1, "group"]  #  "CBC" = 1 = "RED"

data2plot.df$group <- as.factor(data2plot.df$group)
str(data2plot.df)

# 1st eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev1, gwas_ev1)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="1st eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, all variants") +
  geom_abline(slope = 1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(es_evectors.df$es_ev1, wecare_phenotypes.df$eig1_gwas)

# 2nd eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev2, gwas_ev2)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="2nd eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, all variants") +
  geom_abline(slope = -1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(es_evectors.df$es_ev2, wecare_phenotypes.df$eig2_gwas)

# 3rd eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev3, gwas_ev3)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="3rd eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, all variants") +
  geom_abline(slope = -1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(es_evectors.df$es_ev3, wecare_phenotypes.df$eig3_gwas)

# Clean-up
rm(data2plot.df, g)

```

# compare_WES_and_GWAS_eigenvectors_EIGENSTRAT_common_variants

```{r compare_WES_and_GWAS_eigenvectors_EIGENSTRAT_common_variants}

# Prepare plot data

data2plot.df <- as.data.frame(cbind(
  common_es_evectors.df$es_ev1,
  common_es_evectors.df$es_ev2, 
  common_es_evectors.df$es_ev3, 
  wecare_phenotypes.df$eig1_gwas,
  wecare_phenotypes.df$eig2_gwas,
  wecare_phenotypes.df$eig3_gwas,
  wecare_phenotypes.df$cc))

data2plot.df <- cbind(es_evectors.df$wes_id, data2plot.df)

colnames(data2plot.df) <- c("wes_id", "wes_ev1", "wes_ev2", "wes_ev3", 
                            "gwas_ev1", "gwas_ev2", "gwas_ev3", "group")

"UBC" -> data2plot.df[data2plot.df$group == 0, "group"] # "UBC" = 0 = "BLUE"
"CBC" -> data2plot.df[data2plot.df$group == 1, "group"]  #  "CBC" = 1 = "RED"

data2plot.df$group <- as.factor(data2plot.df$group)
str(data2plot.df)

# 1st eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev1, gwas_ev1)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="1st eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, common variants") +
  geom_abline(slope = -1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(common_es_evectors.df$es_ev1, wecare_phenotypes.df$eig1_gwas)

# 2nd eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev2, gwas_ev2)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="2nd eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, common variants") +
  geom_abline(slope = 1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(common_es_evectors.df$es_ev2, wecare_phenotypes.df$eig2_gwas)

# 3rd eigenvectors

g <- ggplot(data2plot.df, aes(wes_ev3, gwas_ev3)) +
  geom_point(aes(colour=group, text=wes_id)) + 
  labs(title="3rd eigenvectors: WES EIGENSTRAT vs GWAS<br>WECARE, all variants") +
  geom_abline(slope = 1, colour = "brown", linetype = 2) + 
  userColourScale
ggplotly(g)

cor.test(common_es_evectors.df$es_ev3, wecare_phenotypes.df$eig3_gwas)

# Clean-up
rm(data2plot.df, g, colours, userColourScale)

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r05d_compare_egenvectors_wecare_only_jan2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
