---
title: "filter_by_variant_effect_wecare_only_jan2017"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 06Mar2017

# Summary

220,812 -> 17,598 variants (8,766 genes)

### Including Loss of function variants
"splice_acceptor_variant", "splice_donor_variant", "stop_gain", "stop_lost", 
"start_lost", "frameshift_variant" : 3,776 variants

### Including dd-missense variants, likely affecting protein function
"deleterious" and "probably_damaging": 13,417 variants (of 82,299 total missenses)

### Including ClinSig variants
"likely_pathogenic", "risk_factor", "pathogenic", "association", "protective", 
"drug_response": 717 variants

### Total retained
**Variants:** 17,598  
**Genes:** 8,766  
~2 protein-affecting variants per gene

## NOT excluding variants with high AF in 1k 

These are sequence versions from Ref genome, which ALTs are common in 1k:

EUR_AF>0.5: 147 from above selection  
EUR_AF>0.9: 19 from above selection (2,092 total in the whole variantset)  

These sequences are preserved in this script.  
However, they will need special treatment in downstream analysis. 

## NOT excluding multiallelic variaints
439 multiallelic varaints in selected  
(5,170 in the whole variantset)

# start_section

```{r start_section}

# Start time
Sys.time()

# Folders
setwd("/scratch/medgen/scripts/wecare_stat_01.17/scripts")
interim_data_folder <- "/scratch/medgen/scripts/wecare_stat_01.17/interim_data"

# Load libraries
library(dplyr)
library(reshape2) # for melt()
library(ggplot2)
library(VennDiagram)

```

# load_data

```{r load_data}

# Load data
load(paste(interim_data_folder, "r06a_exclude_egenvectors_outliers_wecare_only_jan2017.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(wecare_genotypes.mx)
class(wecare_genotypes.mx)
wecare_genotypes.mx[1:5,1:5]

dim(wecare_phenotypes.df)
str(wecare_phenotypes.df)
wecare_phenotypes.df[1:5,1:5]

dim(wecare_variants.df)
colnames(wecare_variants.df)
wecare_variants.df[1:5,1:5]

dim(wecare_kgen.df)
colnames(wecare_kgen.df)
wecare_kgen.df[1:5,1:5]

dim(wecare_exac.df)
colnames(wecare_exac.df)
wecare_exac.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_genotypes.mx) != rownames(wecare_phenotypes.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_variants.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_kgen.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_exac.df))

```

# --- explore_variants_annotations --- #

## explore_VEP_consequences

A combination of the below consequencies can be assigned to a variant

```{r explore_VEP_consequences}

sum(is.na(wecare_variants.df$Consequence)) # 0
a <- strsplit(as.vector(wecare_variants.df$Consequence),"&")
b <- unique(as.vector(unlist(a)))
b
rm(a,b)

```

## explore sift, polyphen, clin_sig

### Annotations seen in wecare

*SIFT:* "tolerated", "deleterious", "tolerated_low_confidence", "deleterious_low_confidence"  

*PolyPhen:* "benign", "possibly_damaging", "probably_damaging", "unknown"  

*CLIN_SIG:* "likely_benign", "benign", "likely_pathogenic", "risk_factor", "not_provided", 
"uncertain_significance", "pathogenic", "association", "protective", "other", "drug_response"  

I have also seen "confers_sensitivity" in CLIN_SIG for gastric cases

### Interpreting SIFT and PolyPhen scores (not used here)
*SIFT:* the higher score the less deleterious  
*PolyPhen:* the higher score the more damaging

```{r explore_sift_polyphen_clinsig}

# SIFT
unique(as.vector(wecare_variants.df$SIFT_call))

# PolyPhen
unique(as.vector(wecare_variants.df$PolyPhen_call))

#CLIN_SIG
unique(unlist(strsplit(as.vector(wecare_variants.df$CLIN_SIG), split="\\&")))

```

## count_variants_frequent_in_kgen_and_exac

2,092 rare sequence versions are present into reference genome (b37)  
~1% of all variants in the datset

Tehy will be preserbved; however they should be treated specially 
in downstream analyses because "variants" called in these sites represent 
sequence versions common in humans, despite the misleading "low" MAFs in some databases. 

```{r count_variants_frequent_in_kgen_and_exac}

x <- wecare_kgen.df  %>% filter(kgen.EUR_AF > 0.9) %>% select(SplitVarID) 
length(x) # 2,092

y <- wecare_exac.df  %>% filter(exac_non_TCGA.AF > 0.9) %>% select(SplitVarID) 
length(y) # 1,497

x <- as.character(x[,1])
y <- as.character(y[,1])
z <- intersect(x,y) 
length(z) # 1,389

rm(x,y,z)

```

## count_multiallelic_varaints

5,170 multiallelic varaints

```{r count_multiallelic_varaints}

summary(wecare_variants.df$Multiallelic) # 5,170

```

# --- Make indices --- #

## make_lof_index

```{r make_lof_index}

sum(is.na(wecare_variants.df$Consequence)) # 0

# --- Splice acceptor variants (sav) --- #

splice_acceptor_variant <- grepl( "splice_acceptor_variant", wecare_variants.df$Consequence)
sum(splice_acceptor_variant) # 317

# mean count of alleles per case in wecare
sav_wecare.mx <- wecare_genotypes.mx[splice_acceptor_variant,]
sav_wecare.counts <- apply(sav_wecare.mx, 2, sum, na.rm=TRUE)
sav_wecare.mean <- mean(sav_wecare.counts, na.rm=TRUE) # ~20
rm(sav_wecare.mx, sav_wecare.counts)

# --- Splice donor variants (sdv) --- #

splice_donor_variant <- grepl( "splice_donor_variant", wecare_variants.df$Consequence)
sum(splice_donor_variant) # 463

# mean count of alleles per case in wecare
sdv_wecare.mx <- wecare_genotypes.mx[splice_donor_variant,]
sdv_wecare.counts <- apply(sdv_wecare.mx, 2, sum, na.rm=TRUE)
sdv_wecare.mean <- mean(sdv_wecare.counts, na.rm=TRUE) # ~17
rm(sdv_wecare.mx, sdv_wecare.counts)

# --- Stop gains --- #

stop_gain <- grepl( "stop_gain", wecare_variants.df$Consequence)
sum(stop_gain) # 1,614

# mean count of alleles per case in wecare
stogv_wecare.mx <- wecare_genotypes.mx[stop_gain,]
stogv_wecare.counts <- apply(stogv_wecare.mx, 2, sum, na.rm=TRUE)
stogv_wecare.mean <- mean(stogv_wecare.counts, na.rm=TRUE) # ~51
rm(stogv_wecare.mx, stogv_wecare.counts)

# --- Stop losts --- #

# Interestingly, there are much less of stop_losts than stop_gains: 56 and 1,168 respectively. 
# This looks contraintuitive, suggesting that stop_gains are better tolerated than stop losses. 
# In other words: stop_losses may be under stronger evolutionary control than stop_gains??

# An alternative explanation coulod be that stop losts can only happen on stop codons; 
# while stop gain can happen in "any"" codon within the gene: so stop gains have much more chances to happen. 

stop_lost <- grepl( "stop_lost", wecare_variants.df$Consequence)
sum(stop_lost) # 56

# mean count of alleles per case in wecare
stolv_wecare.mx <- wecare_genotypes.mx[stop_lost,]
stolv_wecare.counts <- apply(stolv_wecare.mx, 2, sum, na.rm=TRUE)
stolv_wecare.mean <- mean(stolv_wecare.counts, na.rm=TRUE) # ~14
rm(stolv_wecare.mx, stolv_wecare.counts)

# --- Start losts --- #

# Functional significance of start losses may be questioned: a heterozigous start loss may
# be easily compemsated by a mere expression of retained allele?? 

# However, there is not much of them... Either this is an evolutionary control or just "low base" effect (as above for the stop_losses)?

# The total number of start losses is small, so they should not influence the analysis too much anyway ... 

# Why there is no such consequence as "Start gain" ?

start_lost <- grepl( "start_lost", wecare_variants.df$Consequence)
sum(start_lost) # 80

# mean count of alleles per case in wecare
stalv_wecare.mx <- wecare_genotypes.mx[start_lost,]
stalv_wecare.counts <- apply(stalv_wecare.mx, 2, sum, na.rm=TRUE)
stalv_wecare.mean <- mean(stalv_wecare.counts, na.rm=TRUE) # ~7
rm(stalv_wecare.mx, stalv_wecare.counts)

# --- Frameshifts --- #

frameshift_variant <- grepl( "frameshift_variant", wecare_variants.df$Consequence)
sum(frameshift_variant) # 1,277

# mean count of alleles per case in wecare
fsv_wecare.mx <- wecare_genotypes.mx[frameshift_variant,]
fsv_wecare.counts <- apply(fsv_wecare.mx, 2, sum, na.rm=TRUE)
fsv_wecare.mean <- mean(fsv_wecare.counts, na.rm=TRUE) # ~113
rm(fsv_wecare.mx, fsv_wecare.counts)

# --- all LoF-s --- #

lof <- splice_acceptor_variant | splice_donor_variant | stop_gain | stop_lost | start_lost | frameshift_variant

sum(lof) # 3,776

# mean count of alleles per case in wecare
lof_wecare.mx <- wecare_genotypes.mx[lof,]
lof_wecare.counts <- apply(lof_wecare.mx, 2, sum, na.rm=TRUE)
lof_wecare.mean <- mean(lof_wecare.counts, na.rm=TRUE) # ~217
rm(lof_wecare.mx, lof_wecare.counts)

# ------- Plot average counts of lof-s per individual ------- #

# Summary table with average numbers of lof-s per individual
lof_summary_counts.df <- as.data.frame(cbind(
  lof_type=c("splice_acceptor", "splice_donior", "stop_gain", "stop_lost", "start_lost", "frameshift", "all_LoFs"),
  lof_count=c(sav_wecare.mean, sdv_wecare.mean, stogv_wecare.mean, stolv_wecare.mean, stalv_wecare.mean, fsv_wecare.mean, lof_wecare.mean)))

lof_summary_counts.df$lof_count <- as.numeric(as.vector(lof_summary_counts.df$lof_count))
lof_summary_counts.df <- lof_summary_counts.df %>% arrange(desc(lof_count))

# Order of levels defines order of bars on the plot
lof_summary_counts.df$lof_type <- factor(lof_summary_counts.df$lof_type, 
  levels=c("all_LoFs", "frameshift", "stop_gain", "splice_acceptor", "splice_donior", "stop_lost", "start_lost"))

# Plot
ggplot(lof_summary_counts.df, aes(lof_type, lof_count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) +
  geom_text(aes(label=round(lof_count,1)), vjust=-1, size=5)+
  labs(x = "", y = "") + 
  ylim(0,250) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of LoF variants per individual")
  
# Clean-up
rm(splice_acceptor_variant, splice_donor_variant, stop_gain, frameshift_variant, stop_lost,
   start_lost, sav_wecare.mean, sdv_wecare.mean, stogv_wecare.mean, stolv_wecare.mean,
   stalv_wecare.mean, fsv_wecare.mean, lof_summary_counts.df)

```

## make_missense_index

Selecting ~379 missence variants per individual
both: [ deleterious by SIFT ] AND [ probably_damaging by PolyPhen ]

```{r make_missense_index}

missense_all <- grepl( "missense_variant", wecare_variants.df$Consequence)
sum(missense_all) # 83,299

summary(wecare_variants.df$SIFT_call) 
# note many NAs; these are when SIFT is irrelevant (e.g. non-coding)?
deleterious <- wecare_variants.df$SIFT_call == "deleterious"
summary(deleterious)
FALSE -> deleterious[is.na(deleterious)]
summary(deleterious) # 28,388

summary(wecare_variants.df$PolyPhen_call) 
# note many NAs; these are when PolyPhen is irrelevant (e.g. non-coding)?
probably_damaging <- wecare_variants.df$PolyPhen_call == "probably_damaging"
summary(probably_damaging)
FALSE -> probably_damaging[is.na(probably_damaging)]
summary(probably_damaging) # 18,369

missense <- deleterious & probably_damaging
sum(missense) # 13,417

# Plot venn diagram of selected variants by types
grid.newpage()
draw.pairwise.venn(
  area1=sum(deleterious), 
  area2=sum(probably_damaging), 
  cross.area=sum(missense), 
  category=c(
    paste("SIFT\ndeleterious\n", sum(deleterious)),
    paste("PolyPhen\nprobably_damaging\n", sum(probably_damaging))),
  fill=c("red", "blue"), 
  alpha = c(0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

# mean count of missenses per case in wecare
missense_wecare.mx <- wecare_genotypes.mx[missense,]
missense_wecare.counts <- apply(missense_wecare.mx, 2, sum, na.rm=TRUE)
missense_wecare.mean <- mean(missense_wecare.counts, na.rm=TRUE)
missense_wecare.mean # ~379

# Clean-up
rm(missense_all, deleterious, probably_damaging, missense_wecare.mx, missense_wecare.counts)

```

## make_clinsig_index

ClinSig version as reported by Ensembl v87-grch37_vep_cache (Jan2017, VEP script)
ClinVar may become a better variant-annotation database, which may be used in future. 

All *CLIN_SIG* annotations found in the file: "likely_benign", "benign", "likely_pathogenic", "risk_factor", "not_provided", "uncertain_significance", "pathogenic", "association", "protective", "other", "drug_response"  

Note sums w/o na.rm=TRUE.

```{r make_clinsig_index}

selected_clinsig_terms <- c("likely_pathogenic", "risk_factor", "pathogenic", "association", "protective", "drug_response")

clinsig <- wecare_variants.df$CLIN_SIG %in% selected_clinsig_terms
sum(clinsig) # 717

x.df <- as.data.frame(table(as.vector(wecare_variants.df[clinsig,"CLIN_SIG"])))
c("type","count") -> colnames(x.df)
x.df$type <- factor(x.df$type, 
  levels=c("pathogenic", "risk_factor", "likely_pathogenic", "drug_response", "protective", "association"))

ggplot(x.df, aes(type, count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) +
  geom_text(aes(label=round(count,1)), vjust=-1, size=5)+
  labs(x = "", y = "") + 
  ylim(0,600) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Counts of selected ClinSig terms")

# mean count of alleles per case in wecare
clinsig_wecare.mx <- wecare_genotypes.mx[clinsig,]
clinsig_wecare.counts <- apply(clinsig_wecare.mx, 2, sum, na.rm=TRUE)
clinsig_wecare.mean <- mean(clinsig_wecare.counts, na.rm=TRUE) 
clinsig_wecare.mean # 78.8

# Clean-up
rm(selected_clinsig_terms, clinsig_wecare.mx, clinsig_wecare.counts, x.df)

```

# --- explore_selected --- #

## selected_types_per_individual_and_per_variantset

```{r selected_types_per_individual_and_per_variantset}

# Count combinations of selectred categories
lof_missense <- lof & missense
clinsig_missense <- clinsig & missense
lof_clinsig <- lof & clinsig
lof_clinsig_missense <- lof_clinsig & missense

summary(lof_missense)
summary(clinsig_missense)
summary(lof_clinsig)
summary(lof_clinsig_missense)

selected_variants <- lof | clinsig | missense

summary(selected_variants) # 17,598

# --- Plot mean counts of variant types per individual --- #

# Summary table with average numbers of protein-affecting per individual
summary_counts.df <- as.data.frame(cbind(
  type=c("Missense", "LoF", "ClinSig"),
  count=c(missense_wecare.mean, lof_wecare.mean, clinsig_wecare.mean)))

# Convert numbers to numeric vector
summary_counts.df$count <- as.numeric(as.vector(summary_counts.df$count))

# Order of levels defines order of bars on the plot
summary_counts.df$type <- factor(summary_counts.df$type, 
  levels=c("Missense", "LoF", "ClinSig"))

# Plot bar diagramm of avg. count of selected types per individual
ggplot(summary_counts.df, aes(type, count)) +
  geom_bar(stat="identity", fill="blue", alpha=0.5) + 
  geom_text(aes(label=round(count,1)), vjust=-1, size=5)+
  ylim(0, 420) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of selected variants per individual")

# Plot venn diagram of selected variants by types
grid.newpage()
draw.triple.venn(
  area1=sum(lof), 
  area2=sum(clinsig), 
  area3=sum(missense), 
  n12=sum(lof_clinsig), 
  n23=sum(clinsig_missense), 
  n13=sum(lof_missense),
  n123=sum(lof_clinsig_missense),
  category=c(
    paste("lof\n", sum(lof)),
    paste("clinsig\n", sum(clinsig)),
    paste("missense\n", sum(missense))),
  fill = c("red", "green", "blue"), 
  alpha = c(0.3, 0.3, 0.3),
  cex=2, cat.fontface=4, fontfamily=3)

rm(summary_counts.df, missense_wecare.mean, lof_wecare.mean, clinsig_wecare.mean,
   lof, clinsig, missense, lof_clinsig, clinsig_missense, lof_missense, lof_clinsig_missense)

```

# keep_selected_variants_only

220,812 -> 17,598

```{r keep_selected_variants_only}

wecare_genotypes.mx <- wecare_genotypes.mx[selected_variants,]
wecare_variants.df <- wecare_variants.df[selected_variants,]
wecare_kgen.df <- wecare_kgen.df[selected_variants,]
wecare_exac.df <- wecare_exac.df[selected_variants,]

dim(wecare_genotypes.mx)
dim(wecare_variants.df)
dim(wecare_kgen.df)
dim(wecare_exac.df)

# Clean-up
rm(selected_variants)
  
```

# check_frequent_in_Kgen

These variants are preserved;
However, they will require special treatment in downstream analysis

```{r check_frequent_in_Kgen}

x <- wecare_kgen.df$kgen.EUR_AF > 0.9
FALSE -> x[is.na(x)]
x <- wecare_kgen.df[x, "SplitVarID"]
x <- as.character(x)
length(x)
wecare_variants.df[unlist(x),c("SplitVarID","SYMBOL")]

y <- wecare_kgen.df$kgen.EUR_AF > 0.5
sum(y, na.rm=TRUE) # 147

rm(x,y)

```

# count_selected_multiallelic_varaints

439 multiallelics in selected varaints

```{r count_selected_multiallelic_varaints}

summary(wecare_variants.df$Multiallelic) # 439

```

# count_variants_per_gene

```{r count_variants_per_gene}

# Get number of genes
length(unique(wecare_variants.df$SYMBOL)) # 18,766

# Look at the top genes
x <- sort(table(as.vector(wecare_variants.df$SYMBOL)), decreasing = TRUE)
x[1:200]

# Save table of counts
y <- cbind(names(x),x)
c("gene","count") -> colnames(y)
write.table(y, file=paste(results_folder, "wecare_only_variants_per_gene.txt", sep="/"), quote = FALSE, row.names = FALSE)

# Plot histogram
z <- wecare_variants.df %>%  group_by(SYMBOL) %>% summarise(n())
"vars_count" -> colnames(z)[2]
hist(z$vars_count, main="Histogram for numbers of protein-affecting variants per gene", xlab="Variants per gene", ylab="Number of genes", ylim=c(0,7000), breaks=25, labels = TRUE)
mean(z$vars_count) # 2.007529

# Clean-up
rm(x,y,z)

```

# data_summary

```{r data_summary}

dim(wecare_genotypes.mx)
class(wecare_genotypes.mx)
wecare_genotypes.mx[1:5,1:5]

dim(wecare_kgen.df)
colnames(wecare_kgen.df)
wecare_kgen.df[1:5,1:5]

dim(wecare_exac.df)
colnames(wecare_exac.df)
wecare_exac.df[1:5,1:5]

dim(wecare_variants.df)
str(wecare_variants.df)
wecare_variants.df[1:5,1:5]

dim(wecare_phenotypes.df)
str(wecare_phenotypes.df)
wecare_phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(wecare_genotypes.mx) != rownames(wecare_phenotypes.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_kgen.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_exac.df))
sum(rownames(wecare_genotypes.mx) != rownames(wecare_variants.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "s07a_filter_by_variant_effect_wecare_only_jan2017.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
