---
title: "Calculate SKAT for genes, wecare feb2016"
author: "Alexey Larionov"
output: html_document
params:
  interim_data: ""
  subset: ""
  results_folder: ""
  scripts_folder: ""
---

started: Mar2016  
last_updated: 11Aug2016

# Summary

Calculates SKAT for genes

# start_section

```{r start_section}

# Start time
Sys.time()

# Do NOT clean-up at this stage
#rm(list=ls()) # this would remove params object!
graphics.off()

# Read parameters
interim_data_folder <- params$interim_data
data_subset <- params$subset
results_folder <- params$results_folder
scripts_folder <- params$scripts_folder

library(SKAT)

# For debugging
#rm(list=ls())
#interim_data_folder <- "/scratch/medgen/users/alexey/wecare_aug2016/interim_data"
#data_subset <- "priority_genes_strict"
#results_folder <- "/scratch/medgen/users/alexey/wecare_aug2016/results"
#scripts_folder <- "/scratch/medgen/scripts/wecare_skat_08.16/scripts"
#setwd("/scratch/medgen/scripts/wecare_skat_08.16/scripts")

# Check subset
if (! data_subset %in% c("strict", "std.1", "std.2", "relaxed", "priority_genes_relaxed", "priority_genes_strict", "all_variants")){
  print("Wrong data subset")
  stop
}

```

# load_and_check_data

```{r load_and_check_data}

if(data_subset == "strict"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_str_feb2016.RData", sep="/")
}

if(data_subset == "std.1"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_std1_feb2016.RData", sep="/")
}

if(data_subset == "std.2"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_std2_feb2016.RData", sep="/")
}

if(data_subset == "relaxed"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_rel_feb2016.RData", sep="/")
}

if(data_subset == "priority_genes_relaxed"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_pri_rel_feb2016.RData", sep="/")
}

if(data_subset == "priority_genes_strict"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_pri_str_feb2016.RData", sep="/")
}

if(data_subset == "all_variants"){
  source_data_file <- paste(interim_data_folder, "s07_variants_glm_all_feb2016.RData", sep="/")
}

data_subset
source_data_file

# Remove params object to avoid interference with load() function
rm(params)

load(file=source_data_file)
ls()

dim(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(phenotypes.df)
colnames(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(variants.df)
colnames(variants.df)
str(variants.df)
variants.df[1:5,1:5]

dim(variants_assoc.df)
colnames(variants_assoc.df)
str(variants_assoc.df)
variants_assoc.df[1:5,1:5]

# Check consistence of tables
sum(rownames(genotypes.mx) != rownames(variants.df))
sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(rownames(genotypes.mx) != rownames(variants_assoc.df))

```

# load_SKAT_accessory_functions

See separate R file

```{r load_SKAT_accessory_functions, echo=FALSE}

source(paste(scripts_folder,"g01_SKAT_accessory_functions.R", sep="/"))

```

# calculate_associations_per_gene

warning=FALSE  
results='hide'

```{r calculate_associations_per_gene, warning=FALSE, results='hide'}

#---------------------------------------------------#
#                   Report settings                 #
#---------------------------------------------------#

prefix=data_subset
report_type="full"
path=paste(results_folder,"/",sep="")

#---------------------------------------------------#
#                    Phenotypes                     #
#---------------------------------------------------#

phenotypes <- as.vector(phenotypes.df$cc)
names(phenotypes) <- rownames(phenotypes.df)

#---------------------------------------------------#
#         Prepare covariates for the model          #
#---------------------------------------------------#

# --- Tumour --- #

#summary(phenotypes.df$hist_cat)
lobular <- as.numeric(phenotypes.df$hist_cat == "lobular")
#ductal <- as.numeric(phenotypes.df$hist_cat == "ductal")
#medullary <- as.numeric(phenotypes.df$hist_cat == "medullary")
#er1 <- phenotypes.df$er1
stage <- phenotypes.df$stage

# --- Treatment --- #

hormone <- phenotypes.df$hormone == 1
#xray <- phenotypes.df$br_xray_dose
#summary(phenotypes.df$chemo_cat)
#cmf <- as.numeric(phenotypes.df$chemo_cat == "CMF")
#other_cytotoxic <- as.numeric(phenotypes.df$chemo_cat == "Oth")
#no_cytotoxc <- as.numeric(phenotypes.df$chemo_cat == "no")
chemo <- phenotypes.df$chemo_cat != "no"

treatment <- as.numeric(hormone | chemo)

# --- Host --- #

num_preg <- phenotypes.df$num_preg
#BMI_dx <- phenotypes.df$BMI_dx

# --- Etnisity --- #

eigen1 <- phenotypes.df$Eigen_1
eigen2 <- phenotypes.df$Eigen_2
eigen3 <- phenotypes.df$Eigen_3

# --- Complie covariates to numerical matrix --- #

covariates=cbind(lobular, stage, treatment, num_preg, 
                 eigen1, eigen2, eigen3)

colnames(covariates) <- c("lobular", "stage", "treatment", "num_preg", 
                          "eigen1", "eigen2", "eigen3")

rownames(covariates) <- rownames(phenotypes.df)

# --- Clean-up --- #

rm(lobular, stage, hormone, chemo, treatment, num_preg, eigen1, eigen2, eigen3)

#---------------------------------------------------#
#                      Analysis                     #
#---------------------------------------------------#

# Write headers to the results table(s)
prepare_tables_for_results(prefix, report_type, path)

# Get list of genes
genes <- unique(as.vector(variants.df$SYMBOL))

#genes <- c("ATM", "TMCO4", "CHEK1", "CCNL2") 
# ATM is a known candidate, TMCO4 is a random pick
# CHEK1 & CCNL have a single variant per gene, but different allelic frequencies: 
# so they call differnt method.bin and produce different structure of skat test objects

# For some genes (e.g. SLC45A1=gene35 and ROR1=gene323)
# SKATBinary(SKATO) caused crashes with explicit weights:
# e.g. with options [ method="SKATO", weights=c(1,1) ]
# This had to be handled in the accessory functions, 
# see details in comments in the accessory functions file. 

# For each gene
for (gene in genes){
  
  #gene <- genes[3663]
  
  # Get list of variants
  variants_id <- rownames(variants.df[variants.df$SYMBOL == gene,])

  # Initial alleles count in CBC and UBC
  AC_UBC <- sum(as.vector(variants.df[variants_id,"AC_UBC"]), na.rm=TRUE)
  AC_CBC <- sum(as.vector(variants.df[variants_id,"AC_CBC"]), na.rm=TRUE)

  # For genes with a single variant
  if(length(variants_id)==1){
    
    # Get genotypes for the variant
    variant <- genotypes.mx[variants_id,]
    
    # Calculate test
    skat.test <- run_single_variant_test(gene, phenotypes, covariates, variant)
    
    # Print result to report file(s)
    print_results_for_single_variant(skat.test, prefix, AC_UBC, AC_CBC, report_type, path)
    
    
  }else{ # For genes containing multiple variants
    
    # Get the genotypes
    genotypes <- t(genotypes.mx[variants_id,])
    
    # Calculate tests
    skat.test <- run_tests_for_multiple_variants(gene, phenotypes, covariates, genotypes)

    # Print results to report file(s)
    print_results_for_multiple_variants(skat.test, prefix, AC_UBC, AC_CBC, report_type, path)

  }
}

```

# final_section

```{r final_section}

sessionInfo()
Sys.time()
  
```
