---
title: "SKAT_power_estimates: 1000CBC vs 1000UBC"
output: html_document
---

author: AL  
date: 13Feb2017  

This calculation uses data simulated for people of european ancestry.  
It estimates power for a SKAT procedure that aggregates rare variants over a genomic region.  

The simulations below use the following (optimistic?) assumptions:

### N.Sample.ALL = 2480
The total number of samples

### Case.Prop = 0.50
50 % of samples are cases and 50 % of samples are controls  

### Region size = 3kb
~ average size of a gene coding sequence

### Causal.MAF.Cutoff = 5% 
Only SNPs that have MAFs < Causal.MAF.Cutoff may be causal.

###	Causal percent = 50%
Percent of rare SNPs (MAF < Causal.MAF.Cutoff) that are casual.  
This parameter may be set to this relatively high level because we preselect only LOF variants and variants with high SIFT and Polyphen scores.  

###	Negative percent = 0%
Percent of causal variants that have negative effect to the phenotype.  In other words, the below simulation assumes that all variants have effect to the same direction. 

### MaxOR = 2 OR.Type = "Fixed"
Odds ratio for simulated casual variants is fixed = 2 (~ as reported for CHEK2).  

```{r}

# Load SKAT library
library(SKAT)

power_2480 <- Power_Logistic(
  N.Sample.ALL = 2480, 
  Case.Prop=0.5, 
  SubRegion.Length=3000, 
  Causal.MAF.Cutoff=0.05, 
  Causal.Percent= 50, 
  N.Sim=100, 
  MaxOR=2, OR.Type = "Fixed")

power_2480

```

